usaspending_api/disaster/helpers/covid_download_csv_strategies.py

```diff
@@ -0,0 +1,197 @@
+    ):
+        self.spark = None
+        # The place to write intermediate data files to in s3
+        s3_bucket_name = settings.BULK_DOWNLOAD_S3_BUCKET_NAME
+        s3_bucket_path = f"s3a://{s3_bucket_name}"
+        s3_destination_path = f"{s3_bucket_path}/temp_covid_download/{destination_file_name}"
+        try:
+            extra_conf = {
+                # Config for Delta Lake tables and SQL. Need these to keep Dela table metadata in the metastore
+                "spark.sql.extensions": "io.delta.sql.DeltaSparkSessionExtension",
+                "spark.sql.catalog.spark_catalog": "org.apache.spark.sql.delta.catalog.DeltaCatalog",
+                # See comment below about old date and time values cannot be parsed without these
+                "spark.sql.legacy.parquet.datetimeRebaseModeInWrite": "LEGACY",  # for dates at/before 1900
+                "spark.sql.legacy.parquet.int96RebaseModeInWrite": "LEGACY",  # for timestamps at/before 1900
+                "spark.sql.jsonGenerator.ignoreNullFields": "false",  # keep nulls in our json
+            }
+            self.spark = get_active_spark_session()
+            self.spark_created_by_command = False
+            if not self.spark:
+                self.spark_created_by_command = True
+                self.spark = configure_spark_session(**extra_conf, spark_context=self.spark)  # type: SparkSession
+            df = self.spark.sql(source_sql)
+            record_count = write_csv_file(self.spark, df, parts_dir=s3_destination_path, logger=self._logger)
+            # When combining these later, will prepend the extracted header to each resultant file.
+            # The parts therefore must NOT have headers or the headers will show up in the data when combined.
+            header = ",".join([_.name for _ in df.schema.fields])
+            self._logger.info("Concatenating partitioned output files ...")
+            merged_file_paths = hadoop_copy_merge(
+                spark=self.spark,
+                parts_dir=s3_destination_path,
+                header=header,
+                max_rows_per_merged_file=EXCEL_ROW_LIMIT,
+                logger=self._logger,
+            )
+            final_csv_data_file_locations = self._move_data_csv_s3_to_local(
+                s3_bucket_name, merged_file_paths, s3_bucket_path, destination_path
+            )
+        except Exception:
+            self._logger.exception("Exception encountered. See logs")
+            raise
+        finally:
+            delete_s3_object(s3_bucket_name, s3_destination_path)
+            if self.spark_created_by_command:
+                self.spark.stop()
+        append_files_to_zip_file(final_csv_data_file_locations, covid_profile_download_zip_path)
+        self._logger.info(f"Generated the following data csv files {final_csv_data_file_locations}")
+        return final_csv_data_file_locations, record_count
```

Comment:
Maybe there was no other way around this, and it probably works, but it adds a great deal of inefficiency to what will be very large files getting:

1. built in s3
2. downloaded to EC2 filesystem
3. zipped in EC2
4. uploaded to s3

If it's a tradeoff of: 

1. elegantly fitting this new approach of sourcing data from S3 Delta tables (vs. from Postgres) into the existing django command so we can still keep with 1 command and the way that command handled zipping up the fin
2. versus: creating a 2nd (Spark+Delta specific) django command whose execution flow is slightly different than the 1st django command, but allows for more efficient flow to zipping up the final download data

Then my opinion is we should be opting for #2: efficiency over elegance.

To figure out if and how much efficiency is lost, I think we need to see how much time is spent for steps 2 through 4 for the full sized covid files and see if it can't be optimized by performing the Zip _on_ the S3 "FileSystem" using the hadoop copy-merge style code, like what was in the original code:

```
    try:
        # Get the new zipped bits to the target file.
        # If this file is supposed to already exist, there may be a short period of time where the file does NOT
        # exist, between when the delete completes and the rename completes.
        # S3 does not allow renaming to an existing file, and renames are done as copies of the full object
        if overwrite and fs.exists(zip_file_path):
            fs.delete(zip_file_path, True)
        fs.rename(partial_zipped_file_path, zip_file_path)
    except Exception:
        if fs.exists(partial_zipped_file_path):
            fs.delete(partial_zipped_file_path, True)  # drop partial zip file if rename completes or not
        raise

    logger.info(
        f"Completed zip of {len(merged_file_paths)} {extension} files into file "
        f"{str(zip_file_path)} in {(time.time() - zip_start):3f}s"
    )
```

e.g. given the size of the files, I think it'd be ideal if we could build and stage them all in S3, and perform the Zip _right in S3_ as the above code does, avoiding the download and upload.

Reply:
Discussed in PR review and likely to push investigation of efficiency to next Story in this series, as a possible improvement.